### PR TITLE
kvstorage: complete RaftReplicaID migration

### PR DESCRIPTION
**Pull request by tbg**
As of v22.1[^1], we always write the RaftReplicaID when creating a
Replica or updating a snapshot. However, since this is
persisted state that could've originated in older versions and not
updated yet, we couldn't rely on a persisted ReplicaID yet.

This commit adds code to the `(*Store).Start` boot sequence that

- persists a RaftReplicaID for all initialized replicas (using the
  ReplicaID from the descriptor)
- deletes all uninitialized replicas lacking RaftReplicaID (since we don't know their
  ReplicaID at this point).

The second item in theory violates Raft invariants, as uninitialized
Replicas are allowed to vote (though they then cannot accept log
entries). So in theory:

- an uninitialized replica casts a decisive vote for a leader
- it restarts
- code in this commit removes the uninited replica (and its vote)
- delayed MsgVote from another leader arrives
- it casts another vote for the same term for a dueling leader
- now there are two leaders in the same term.

The above in addition presupposes that the two leaders cannot
communicate with each other. Also, even if that is the case, since the
two leaders cannot append to the uninitialized replica (it doesn't
accept entries), we also need additional voters to return at the exact
right time.

Since an uninitialized replica without RaftReplicaID in is necessarily
at least one release old, this is exceedingly unlikely and we will
live with this theoretical risk.

This PR also adds a first stab at a datadriven test harness for
`kvstorage` which is likely to be of use for #93247.

[^1]: https://github.com/cockroachdb/cockroach/pull/75761

Epic: [CRDB-220](https://cockroachlabs.atlassian.net/browse/CRDB-220)
Release note: None

### CODE REVIEW

**Comment by tbg** [comment]
Now it's good to review.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`-- commits` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95513#-NMUpUs4BRdF5TQA6s09):*
wasn't it v22.1?

There is also a similar comment in https://github.com/cockroachdb/cockroach/blob/c9f87cf306f7b23835e5d2aa7edf2c2ad38b66dd/pkg/kv/kvserver/store_create_replica.go#L249-L283 which could use some adjustment in light of this PR doing this Replica cleanup.

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @sumeerbhola)

___
*[`-- commits` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95513#-NMUpUs4BRdF5TQA6s09):*
> wasn't it v22.1?

Yep, thanks.

Updated the comment.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 1]
```diff
@@ -347,6 +375,20 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 					ident.StoreID, desc)
 			}
 
+			sp := desc.RSpan().AsRawSpanWithNoLocals()
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:378]: Could you add a comment above saying what this block does? It looks to me like it checks that all range descriptors are non-overlapping.

Do you need the interval tree here? If `IterateRangeDescriptorsFromDisk` returns descriptors ordered by the first key (and the `INVARIANT` also says that begin/end keys are verified to match the key at which the descriptor is located), you only need to check that `current.Key` >= `prev.EndKey`. You can also double-check that `current.Key` > `prev.Key` (unless you trust the iteration func, in which case it's sufficient to make a comment).

Side: this seems out of place in this commit/PR, would prefer to have reviewed it separately.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 2]
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
+	//
+	// This invariant is true for replicas created in 22.2. Without further action, it
+	// maybe be violated for clusters that originated before 22.2.
+	//
+	// loadFullReplicaIDsFromDisk returns a zero ReplicaID when none is found.
+	//
+	// In this method, we migrate to always satisfy the invariant:
+	//
+	// 1. for initialized replicas, use ReplicaID from RangeDescriptor as RaftReplicaID
+	// 2. remove uninitialized replicas lacking a RaftReplicaID.
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:411]: (2) doesn't look like invariant. Did you mean to say that "uninitialized replicas don't have a stored RaftReplicaID"? This isn't yet true either, because `getOrCreateReplica` still writes one asap.

Should we hold off deleting uninit replica IDs from storage until (or after) a PR that would stop writing them too? Otherwise the code will be in a state where we always write them, and remove on restarts, which is a bit wasteful.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 3]
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
+	for id := range allM {
+		if id.ReplicaID != 0 {
+			// RaftReplicaID was present, no need to backfill.
+			continue
 		}
-		// `allM` will be our map of uninitialized replicas.
+
+	}
+
+	for id := range initM {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:430]: Not reviewed today below this line. Will continue tomorrow, but also would like to see if this PR can be split: (1) complete backfilling ReplicaIDs, (2) some time later also look at removing ReplicaIDs that we don't need (rationale in the prev. comment).

File: pkg/kv/kvserver/kvstorage/init.go (hunk 2 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/kvstorage/init.go:411]: I misunderstood the comment. (1) and (2) describe migration actions, which are needed to arrive at the invariant above. I might be wrong in my interpretation of what this does, will take a fresh look tomorrow.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/kvstorage/init.go:411]: Ok, so we're looking at 3 KVs of a `Replica`: `RaftReplicaID`, `HardState`, and `RangeDescriptor`.

<table>
<tr>
	<td> RaftReplicaID
	<td> HardState
	<td> RangeDescriptor
	<td> What?
<tr>
	<td> no
	<td> no
	<td> no
	<td> nothing
<tr>
	<td> no
	<td> no
	<td> yes
	<td> -
<tr>
	<td> no
	<td> yes
	<td> no
	<td> uninitialized (v <= 22.1)
<tr>
	<td> no
	<td> yes
	<td> yes
	<td> initialized (v <= 22.1)
<tr>
	<td> yes
	<td> no
	<td> no
	<td> -
<tr>
	<td> yes
	<td> no
	<td> yes
	<td> -
<tr>
	<td> yes
	<td> yes
	<td> no
	<td> uninitialized (v >= 22.2)
<tr>
	<td> yes
	<td> yes
	<td> yes
	<td> initialized
</table>

Steps 1 and 2 in this comment say that we eliminate the "RaftReplicaID: no" half of this table. The way step 1 and 2 do so is different though, because for uninit replica we don't know the Replica ID; if we did we would write it too; but instead we remove such replicas.

Makes sense.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/kvstorage/init.go:411]: It's a bit hard to follow all the map maths below. I will try now, but I think this could be simpler if we just implemented something like a table that a drew above:

```golang
type rangeState struct {
	replicaID roachpb.ReplicaID
	hardState raftpb.HardState
	rangeDesc roachpb.RangeDescriptor
}

var m map[roachpb.RangeID]rangeState

read all RaftReplicaID
insert into m[rangeID].replicaID (init a new rangeState if it does not exist)

read all HardState
insert into m[rangeID].hardState (init a new rangeState if it does not exist)

read all RangeDescriptor
insert into m[rangeID].rangeDesc (init a new rangeState if it does not exist)

for rangeID, state := range m {
	checkInvariants(rangeID, state) // e.g. assert that we don't see one of the impossible "-" combinations
	writeMigration(rangeID, state)
}
```

File: pkg/kv/kvserver/kvstorage/init.go [hunk 4]
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
-			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+	for id := range allM {
+		if id.ReplicaID != 0 {
+			// RaftReplicaID was present, no need to backfill.
+			continue
 		}
-		// `allM` will be our map of uninitialized replicas.
+
+	}
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:428]: This loop looks like a no-op? Or maybe it's some indentation that confused me.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 2 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go]: Good idea, this simplified it a lot. I'll ping you when the review is ready.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 3 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go]: I rewrote according to your suggestion and I think the result is orders of magnitude clearer. I tend to think splitting it at this point is more busy work for me than benefit for you but let me know (once I've pinged) and I can still tease it apart after.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go]: I brought it back, since it's a one-liner now. I had it in a separate commit but the whitespace changes made it more trouble than is worth it.

**Comment by tbg** [comment]
Ok, RFAL.

**Comment by tbg** [comment]
https://teamcity.cockroachdb.com/repository/download/Cockroach_Ci_Tests_Testrace/8563775:id/bazel-testlogs/pkg/storage/storage_test/test.log

```
panic: concurrent write operations detected on file [recovered]
	panic: concurrent write operations detected on file

goroutine 12121 [running]:
github.com/cockroachdb/pebble.(*DB).runCompaction.func1()
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:2258 +0x23d
panic({0x27ac820, 0x31d3a20})
	GOROOT/src/runtime/panic.go:890 +0x262
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).timeDiskOp(0xc000584910, 0x2, 0xc000988558)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:262 +0x1a5
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).Sync(0xc000584910)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:219 +0x69
github.com/cockroachdb/pebble/vfs.(*enospcFile).Sync(0xc000138648)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:391 +0x6d
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc00090a500, 0x7d, 0xc000539200)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:2836 +0x312f
github.com/cockroachdb/pebble.(*DB).flush1(0xc00090a500)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1685 +0x61d
github.com/cockroachdb/pebble.(*DB).flush.func1({0x31f15b8, 0xc000a0f2c0})
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1624 +0x137
runtime/pprof.Do({0x31f1548, 0xc0000600a8}, {{0xc0000acf60?, 0xc00090a500?, 0x0?}}, 0xc00073a7a0)
	GOROOT/src/runtime/pprof/runtime.go:40 +0x123
github.com/cockroachdb/pebble.(*DB).flush(0xc00090a500)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1617 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1533 +0x186
I230202 22:16:32.090098 1 (gostd) testmain.go:626  [T1] 1  Test //pkg/storage:storage_test exited with error code 2
```

**Comment by knz** [comment]
@RaduBerinde requested that we file these concurrency issues inside pebble as issues on the pebble repo.

**Comment by tbg** [comment]
Filed https://github.com/cockroachdb/pebble/issues/2301

File: pkg/kv/kvserver/kvstorage/testdata/init [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Uninitialized deprecated replica: doesn't have a RaftReplicaID, it's just a
+# HardState. We expect this to be removed by load-and-reconcile.
+new-replica range-id=5
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/testdata/init:3]: r5 doesn't show up in any way in the output below, except in comments that reason about what happened to it. Is there a way to surface it in the output? Or is the absence of r5 in the output the intended signal?

File: pkg/kv/kvserver/kvstorage/init.go [hunk 6]
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
...
 }
 
-type EngineReplicas struct {
-	Uninitialized map[storage.FullReplicaID]struct{}
-	Initialized   map[storage.FullReplicaID]*roachpb.RangeDescriptor
+// LoadedReplicas represents the Replicas present on a storage engine.
+type LoadedReplicas struct {
+	Uninitialized ReplicaMap
+	Initialized   ReplicaMap
 }
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:300]: IMO, this type is not needed, one can return `ReplicaMap` directly and sift through (un)inited replicas if needed.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 7]
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
...
-		return nil
-	}); err != nil {
-		return nil, err
+// A Replica references a CockroachDB Replica. The data in this struct does not
+// represent the data of the Replica but is sufficient to access all of its
+// contents via additional calls to the storage engine.
+type Replica struct {
+	RangeID   roachpb.RangeID
+	ReplicaID roachpb.ReplicaID
+	Desc      *roachpb.RangeDescriptor // nil for uninitialized Replica
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:308]: Would a default-constructed `RangeDescriptor` suffice (treat `Desc.RangeID == 0` as nil), or you are avoiding some overhead here by making it a pointer?

File: pkg/kv/kvserver/kvstorage/init.go [hunk 8]
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
...
 	}
+}
+
+// A ReplicaMap organizes a set of Replicas with unique RangeIDs.
+type ReplicaMap map[roachpb.RangeID]Replica
 
-	// TODO(sep-raft-log): if there is any other data that we mandate is present here
-	// (like a HardState), validate that here.
+// SortedSlice returns the contents of the map as a RangeID-sorted slice.
+func (m ReplicaMap) SortedSlice() []Replica {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:326]: nit: I'd name it `Sorted()`. The "slice" bit is obvious.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 9]
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
...
+	var sl []Replica
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:327]: Pre-allocate for `len(m)`?

File: pkg/kv/kvserver/kvstorage/init.go [hunk 10]
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
...
+	ent.Desc = &desc
+	m[rangeID] = ent
+	return nil
+}
+
+func (m ReplicaMap) setDescReplicaID(rangeID roachpb.RangeID, descReplicaID roachpb.ReplicaID) {
+	ent := m.getOrMake(rangeID)
+	ent.descReplicaID = descReplicaID
+	m[rangeID] = ent
+}
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:369]: Why is this needed? Could we check the invariant using `Desc -> ... -> ReplicaID` directly instead of pre-computing / denormalizing it when loading?

File: pkg/kv/kvserver/kvstorage/init.go [hunk 11]
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 					ident.StoreID, desc)
 			}
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go]: I think we could push the invariant check down where you check other invariants. This would allow removing `Replica.descReplicaID` and `setDescReplicaID`.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 12]
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
 
-	allM, err := loadFullReplicaIDsFromDisk(ctx, eng)
-	if err != nil {
-		return nil, err
+	// INVARIANT: all replicas have a persisted full ReplicaID (i.e. a "ReplicaID from disk").
+	//
+	// This invariant is true for replicas created in 22.1. Without further action, it
+	// would be violated for clusters that originated before 22.1. In this method, we
+	// backfill the ReplicaID (for initialized replicas) and we remove uninitialized
+	// replicas (see below for rationale).
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:421]: Removing all uninitialized replicas? Or the ones that don't have a `RaftReplicaID`? Worth clarifying here and in the commit description.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 13]
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
+			continue
+		}
+		if repl.descReplicaID != 0 {
+			// Backfill RaftReplicaID for an initialized Replica.
+			if err := logstore.NewStateLoader(repl.RangeID).SetRaftReplicaID(ctx, eng, repl.descReplicaID); err != nil {
+				return LoadedReplicas{}, errors.Wrapf(err, "backfilling ReplicaID for r%d", repl.RangeID)
 			}
-			if err := logstore.NewStateLoader(id.RangeID).SetRaftReplicaID(ctx, eng, id.ReplicaID); err != nil {
-				return nil, errors.Wrapf(err, "backfilling replicaID for r%d", id.RangeID)
+			s.setReplicaID(repl.RangeID, repl.descReplicaID)
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:473]: This `set` modifies Go `map` under the hood. I wonder how that impacts the iteration on it, do you know? Maybe a more transparent approach would be:

```golang
repls := s.SortedSlice()
for i, repl := range repls {
	repls[i].foo = bar
}
```

UPD: aah, we're iterating the slice, not the map.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 14]
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
+			log.Eventf(ctx, "backfilled replicaID for initialized replica %s", s.getOrMake(repl.RangeID).ID())
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:474]: nit: `getOrMake` seems unnecessary here, could construct `ID` directly.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 15]
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
...
-		// A replica is "uninitialized" if it's not in initM (i.e. is at log position
-		// zero and has no visible RangeDescriptor).
-		delete(allM, id)
 	}
 
-	return &EngineReplicas{
-		Initialized:   initM,
-		Uninitialized: allM, // NB: init'ed ones were deleted earlier
-	}, nil
+	return LoadedReplicas{Initialized: init, Uninitialized: uninit}, nil
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:503]: `return s`?

File: pkg/kv/kvserver/store_create_replica.go [hunk 16]
```diff
@@ -245,42 +245,19 @@ func (s *Store) tryGetOrCreateReplica(
 	//   this newer replica is harmless since it just limits the votes for
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/store_create_replica.go:245]: unrelated drive-by question: Is "earlier replica" possible here? Why only "same" or "later"?

**Review by pav-kv — APPROVED** [review]
Looking much clearer now, thanks for implementing the suggestion! LGTM in principle, but left some tips/nits.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 13 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/kvstorage/init.go:473]: Since we've sorted it already either way, I would return the sorted slice from the func, so that the caller doesn't need to sort it again.

Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:473]: Also, we're updating an existing element in the map, so it's fine. If you're adding elements or removing while iterating, the element may or may not show up.

File: pkg/kv/kvserver/kvstorage/testdata/init (hunk 5 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/testdata/init:3]: It does show up:

image

Basically it was tossed out.

File: pkg/kv/kvserver/kvstorage/datadriven_test.go [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+					fmt.Fprintln(&buf, err)
+					break
+				}
+				var merged []storage.FullReplicaID
+				for id := range rs.Initialized {
+					merged = append(merged, id)
+				}
+				for id := range rs.Uninitialized {
+					merged = append(merged, id)
+				}
```
[the remark this answers is not in the document]
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/datadriven_test.go:175]: Good suggestion, that simplified things.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 6 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:300]: Done!

File: pkg/kv/kvserver/kvstorage/init.go (hunk 7 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:308]: We could use the zero value here, but what's the upshot? I know there's the Go koan of "making the zero value useful" but a `RangeDescriptor` can not be compared to its zero value (since it contains a slice) and besides it's much more noisy than just checking if a pointer is nil. For better or worse a nil pointer is the closest we got to a rust `Option` type (maybe generics help 👿 ).

File: pkg/kv/kvserver/kvstorage/init.go (hunk 12 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:421]: Done.

Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:421]: (in comment, commit, PR)

File: pkg/kv/kvserver/store_create_replica.go (hunk 16 above)
Comment (tbg) [reply pkg/kv/kvserver/store_create_replica.go:245]: Earlier might be possible. I think there is a hole here, the tombstone and the RaftReplicaID must really work in unison - if RaftReplicaID is 10 then a tombstone of 5 should be treated as a tombstone of 9.

Let's discuss on Slack, I'd like to get this PR in :-)

**Comment by tbg** [comment]
Still have to push the updates, please ignore my comment pings until I request.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 10 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:369]: Good idea, doing this.

File: pkg/kv/kvserver/kvstorage/testdata/init [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+# Loading the replicas returns only the ones that
+# had a ReplicaID.
+load-and-reconcile trace=true
+----
+r6/60: uninitialized
+r7/70: r7:{a-c} [(n1,s1):70, next=71, gen=0]
+r8/80: r8:{c-f} [(n1,s1):80, next=81, gen=0]
+beginning range descriptor iteration
+iterated over 2 keys to find 2 range descriptors (by suffix: map[rdsc:2])
+removed legacy uninitialized replica for r5
```
Comment (tbg) [inline pkg/kv/kvserver/kvstorage/testdata/init:31]: Here, @pavelkalinnikov

File: pkg/kv/kvserver/kvstorage/init.go [hunk 19]
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
...
+			ctx, eng, func(desc roachpb.RangeDescriptor) error {
+				if lastDesc.RangeID != 0 && desc.StartKey.Less(lastDesc.EndKey) {
+					return errors.AssertionFailedf("overlapping descriptors %s and %s", lastDesc, desc)
+				}
+				lastDesc = desc
+
+				if err := s.setDesc(desc.RangeID, desc); err != nil {
+					return err
+				}
+				return nil
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:388]: supernit: `return s.setDesc(...)`

File: pkg/kv/kvserver/kvstorage/init.go [hunk 20]
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
...
+			}); err != nil {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:389]: supernit: I'd put comma here and `); err != nil` check on the next line to avoid the weird indentation.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 21]
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
...
+	ent := m[rangeID]
+	ent.RangeID = rangeID
+	return ent
+}
+
+func (m ReplicaMap) setReplicaID(rangeID roachpb.RangeID, replicaID roachpb.ReplicaID) Replica {
+	ent := m.getOrMake(rangeID)
+	ent.ReplicaID = replicaID
+	m[rangeID] = ent
+	return ent
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:340]: nil: dunno, this is a bit weird, can we avoid this API?

I think (idea 1):
- The user of `LoadAndReconcileReplicas` doesn't need the map, so it can return `[]Replica` directly.
- Since we're sorting anyway (for determinism), let it return a sorted `[]Replica`.
- Then, the tests don't need to `Sorted()` the map again.
- Also, `ReplicaMap` and all its methods (incl. `Sorted`) can be unexported.
- Inside `LoadAndReconcileReplicas` in the place where we start overriding things in the map, we should just convert to sorted slice, and modify slice elements directly.

Idea 2: If you really want to expose the map-ness to the outside (e.g. the user of this needs to search by RangeID, which though it currently doesn't), I could also imagine a data structure like this:

```golang
type ReplicaMap struct {
	index map[roachpb.RangeID]int
	repls []Replica
}

func (m *ReplicaMap) getOrMake(rangeID roachpb.RangeID) int { // or *Replica
	if index, found := m[rangeID]; found {
		return index
	}
	m.repls = append(m.repls, Replica{RangeID: rangeID})
	return len(m.repls) - 1
}

func (m *ReplicaMap) Sort() { // or stateless Sorted() like right now
	// sort m.repls
	// reindex m.index using m[].RangeID as key
}

// then all the accessors are not needed, can just `getOrMake` and write
// to the relevant returned *Replica (or repls[index]) directly, like this:
// 
m.getOrMake(rangeID).hardState = hs
// or
m.repls[m.getOrMake(rangeID)].hardState = hs
```

Some combination of idea 1 and 2 would be nice.

**Review by pav-kv — APPROVED** [review]
Some final bits and bobs.

File: pkg/kv/kvserver/kvstorage/init.go (hunk 21 above)
Comment (tbg) [reply pkg/kv/kvserver/kvstorage/init.go:340]: Went with 1), the yak has been shaved 👑 
I like the result better.

File: pkg/kv/kvserver/kvstorage/init.go [hunk 22]
```diff
@@ -291,103 +293,198 @@ func IterateRangeDescriptorsFromDisk(
...
+			//
+			// TODO(tbg): if clearRangeData were in this package we could destroy more
+			// effectively even if for some reason we had in the past written state
+			// other than the HardState here (not supposed to happen, but still).
+			if err := eng.ClearUnversioned(logstore.NewStateLoader(repl.RangeID).RaftHardStateKey()); err != nil {
+				return nil, errors.Wrapf(err, "removing HardState for r%d", repl.RangeID)
 			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+			log.Eventf(ctx, "removed legacy uninitialized replica for r%s", repl.RangeID)
+			sl = append(sl[:idx], sl[idx+1:]...)
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/kvstorage/init.go:483]: To avoid the O(N * M) complexity here, in a case when N out of M replicas are legacy uninited replicas, could instead rewrite one `Replica` in the slice at a time:

```golang
back := 0
for idx, repl := range sl {
	...
	if !(remove condition) {
		sl[back] = repl
		back++
	}
}
return sl[:back], nil
```

May require a bit of shuffling/reordering to work around the `continue`.

**Review by pav-kv — APPROVED** [review]

**Comment by tbg** [comment]
bors r=pavelkalinnikov
